### PR TITLE
Previews: Fix for previewing posts on sites with mapped domains.

### LINKED ISSUE #14220: Editor: cannot preview drafts on a WordPress.com site with a custom domain
We received the following report:

> I am currently unable to use the View and Preview functions on my Wordpress App on iPad. I receive a 404 page not found error each time. The issue arises when I attempt to preview or view my drafts. Any of the drafts on my site will show the effect.

* Type of site: WordPress.com Premium with a custom domain
* WPiOS 14.8.2
* iPad Pro 9.7-inch (WiFi/Cellular) (iPad6,4)
* iOS 13.4.1

Note: the app logs provided by the user showing an editor session with a preview doesn't seem to indicate a problem (is it possible we aren't logging failed previews in this flow and so app logs won't be helpful in this type of case?):

```
2020-05-25 09:58:46:344 🔵 Tracked: editor_session_start <blog_type: wpcom, content_type: gutenberg, editor: gutenberg, has_unsupported_blocks: 0, post_type: post, session_id: 218FF820-7CD8-4A57-AB4E-1D4949111147, startup_time_ms: 1723, unsupported_blocks: (
)>
2020-05-25 09:58:58:388 🔵 Tracked: web_preview_opened <blog_type: wpcom, post_type: post>
2020-05-25 09:59:00:339 TracksService sendQueuedEvents completed. Sent 2 events.
2020-05-25 09:59:03:718 🔵 Tracked: editor_session_end <blog_type: wpcom, content_type: gutenberg, editor: gutenberg, has_unsupported_blocks: 0, outcome: cancel, post_type: post, session_id: 218FF820-7CD8-4A57-AB4E-1D4949111147>
2020-05-25 09:59:03:728 🔵 Tracked: editor_discarded_changes <blog_id: 151743515, editor_source: gutenberg, has_gutenberg_blocks: 1, post_id: 647>
2020-05-25 09:59:03:780 🔵 Tracked: editor_closed <blog_id: 151743515, editor_source: gutenberg, has_gutenberg_blocks: 1, post_id: 647>
```

<p style="clear:both;"><sup>(internal reference: p4a5px-2zK-p2, 2987505-zen)</sup></p>

### PR DESCRIPTION
**Pull request by aerych**
Fixes #14220 

This PR adds new logic to preview URLs to ensure that previewing on a site with a mapped domain will use the site's unmapped URL rather than the mapped URL.  This will let our preview browser use existing auth cookies rather than having to make a complicated series of requests to obtain an authenticated session.

@ScoutHarris could I trouble you for a review?  Also ccing @guarani who has looked into this previously.

** Merge after landing related changes to WordPressKit.

To test:
Prep:
- Start from a fresh install so there are no cookies lingering in the cookie jar.   For an extra level of paranoia, delete and reinstall for each scenario.

Scenario 1: Mapped Domain Draft Preview
- [ ] Create a new draft post on a wpcom site with a mapped domain. 
- [ ] Preview the draft. 
- [ ] Confirm you see the remote preview and not an error.

Scenario 2: Mapped Domain Published Preview
- [ ] Edit an already published post on a wpcom site with a mapped domain. 
- [ ] Preview your changes. 
- [ ] Confirm you see the remote preview and not an error.

Scenario 3: Unmapped Site Draft Preview
- [ ] Repeat Scenario 1 for this site type. Confirm there is no regression.

Scenario 4: Unmapped Published Preview
- [ ] Repeat Scenario 2 for this site type. Confirm there is no regression.

Scenario 5: Jetpack Site Draft Preview
- [ ] Repeat Scenario 1 for this site type. Confirm there is no regression.

Scenario 6: Jeptack Site Published Preview
- [ ] Repeat Scenario 2 for this site type. Confirm there is no regression.

Scenario 7: Self-hosted Site Draft Preview
- [ ] Repeat Scenario 1 for this site type. Confirm there is no regression.

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/Preview/PreviewNonceHandler.swift [hunk 1]
```diff
@@ -22,4 +25,26 @@ struct PreviewNonceHandler {
         components.queryItems = queryItems
         return components.url
     }
+
+    private static func unmapURL(post: AbstractPost, url: URL) -> URL? {
+        guard
+            let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
```
Comment (ScoutHarris) [inline WordPress/Classes/ViewRelated/Post/Preview/PreviewNonceHandler.swift]: Suggestion - if this was a `var` you wouldn't need `comps` below.

**Comment by ScoutHarris** [comment]
Hey @aerych .

Here are my testing results.

**Successes:**
✅ Scenario 1: Mapped Domain Draft Preview 
✅ Scenario 3: Unmapped Site Draft Preview
✅ Scenario 4: Unmapped Published Preview
✅ Scenario 5: Jetpack Site Draft Preview
✅ Scenario 7: Self-hosted Site Draft Preview


**Failures:**
🚫 Scenario 2: Mapped Domain Published Preview - the preview doesn't show edits until after the post is saved.

Both of these behaved the same:
🚫 Scenario 6: Jetpack Site Published Preview - if it matters, I was using a self-hosted site with Jetpack installed. (via jurassic.ninja)
🚫 Not mentioned: Self-hosted Published Preview - I was using a self-hosted site without Jetpack. (via jurassic.ninja)

- On a published post > Edit > Preview > shows preview successfully.
- On a published post > Edit > edit the post > Preview > shows the `Preview Unavailable` error. Also, this was in the logs:

`Error while trying to save post before preview: Optional(Error Domain=PostServiceErrorDomain Code=0 "Previews are unavailable for this kind of post." UserInfo={NSLocalizedDescription=Previews are unavailable for this kind of post.})`

**Review by ScoutHarris — CHANGES_REQUESTED** [review]
See testing results.

File: WordPress/Classes/ViewRelated/Post/Preview/PreviewNonceHandler.swift (hunk 1 above)
Comment (aerych) [reply WordPress/Classes/ViewRelated/Post/Preview/PreviewNonceHandler.swift]: Oh ha! I never realized `guard var` was allowed! TIL.  Thanks!

**Comment by aerych** [comment]
Thank you @ScoutHarris !

> Scenario 2: Mapped Domain Published Preview - the preview doesn't show edits until after the post is saved

For this one I found two scenarios where the preview wouldn't show.  One was when proxied 😬.  The other depended on the site's permalink format.  I've pushed an update that should take care of this case. Would you mind trying again?

> Scenario 6: Jetpack Site Published Preview - if it matters, I was using a self-hosted site with Jetpack installed. (via jurassic.ninja)

Arg! I even [had a recording](https://github.com/wordpress-mobile/WordPress-iOS/issues/11684#issuecomment-704665842) of this working. :(   Sadly, it looks like I was [testing this under special circumstances](https://github.com/wordpress-mobile/WordPress-iOS/issues/11684#issuecomment-705207599) so my expectations were off.   The preview is _not_ expected to be shown in this case so, what you saw was probably expected and not a regression, but please confirm!

 > Not mentioned: Self-hosted Published Preview - I was using a self-hosted site without Jetpack

Yeah. I omitted this because these previews were not expected to work in this scenario. (No autosave post type.) :)

Ready for another peek!

**Comment by ScoutHarris** [comment]
Hey @aerych .

> For this one I found two scenarios where the preview wouldn't show. One was when proxied 😬. The other depended on the site's permalink format. I've pushed an update that should take care of this case. Would you mind trying again?

I'm not sure what's wrong but my .blog site doesn't show unsaved edits in the preview. I tried on a real device, so I wasn't proxied.

| post_edit | post_preview |
|--------|-------|

**Comment by aerych** [comment]
Hmm... I wonder if we might be taking different steps?  This is what I'm doing:
preview-published
Lemme know if you spot a difference and I'll try it and see if I can repo what you're seeing.

**Comment by aerych** [comment]
Just in case using a device mattered, I repeated the above test on my iPhone 11 Pro (iOS 13.7) and I'm seeing the same result. (I did use different test text just to be sure.) The previous test was in the iPhone 11 Pro (iOS 14) simulator. 

I'm wondering if the issue you're seeing might be due to a glitch with the cookie auth (not sure yet how to test this).  I can look into this angle tomorrow.

**Comment by aerych** [comment]
That was a merry little debugging adventure :) 

So happy you happened to test with an Atomic site!  After logging in to your test site as an editor (thank you again!) I was able to reproduce the issue there.   To recap our chat conversation, the issue turned out to be that for Atomic sites, the value of their `unmapped_url` is still their mapped domain name, not a wpcom subdomain like for my test site.  This meant that the wpcom auth cookies weren't valid for the atomic site, so the published version of the post was shown rather than the autosaved version. This makes sense considering the way Atomic sites are configured.

Some investigation revealed the trick Calypso uses is to load wp-login.php with a redirect to the desired page.  In the case of Atomic sites with SSO enabled the redirect is fairly seamless.  In cases where SSO is not enabled, there is still a prompt for credentials as a semi-graceful fallback.  It's not perfect, but better than the current situation. 

I've updated the PR with some logic that checks for public atomic sites and rewrite the request URL to proxy through wp-login.php. I tested with your test site and it was working for me there.  I also created an ephemeral atomic site to test with but, interestingly, these sites do not report themselves as being atomic (?!) via the REST API and so regular wpcom auth is attempted.  This is something that needs to be addressed server-side, not in the app, so 🤷 for now.  I'll open an Android issue regarding Atomic published post previews. 

Ready for another peek!

cc @diegoreymendez in case you're interested in this as well.

**Review by ScoutHarris — APPROVED** [review]
Excellent debugging skillz!! It is in fact working as advertised. Well done!

:shipit:

File: WordPress/Classes/Utility/Networking/RequestAuthenticator.swift [hunk 2]
```diff
@@ -182,6 +191,36 @@ class RequestAuthenticator: NSObject {
...
+            queryItems.append(URLQueryItem(name: "redirect_to", value: url.absoluteString))
+            components?.queryItems = queryItems
+            let requestURL = components?.url ?? url
+
+            let request = URLRequest(url: requestURL)
+            completion(request)
+        }
+
+        authenticationService.loadAuthCookiesForWPCom(into: cookieJar, username: username, authToken: authToken, success: {
+            done()
```
Comment (diegoreymendez) [inline WordPress/Classes/Utility/Networking/RequestAuthenticator.swift:211]: Fairly minor, but this jumped out at me while looking at the code.  I think you can simplify the success param to just be:

```swift
success: done,
```

(Without the `{}`s and `()`s)

Comment (aerych) [reply WordPress/Classes/Utility/Networking/RequestAuthenticator.swift:211]: True. If you don't mind I think I'll leave it as is so it's consistent with the other methods in that class.   Thanks for taking a peek!

**Comment by aerych** [comment]
Thank you @ScoutHarris and @diegoreymendez !

**Pull request merged**